File: bot/engine/src/main/kotlin/fr/vsct/tock/bot/engine/dialog/Story.kt
/*
 * Copyright (C) 2017 VSCT
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package fr.vsct.tock.bot.engine.dialog

import fr.vsct.tock.bot.definition.Intent
import fr.vsct.tock.bot.definition.StoryDefinition
import fr.vsct.tock.bot.definition.StoryHandler
import fr.vsct.tock.bot.engine.BotBus
import fr.vsct.tock.bot.engine.BotRepository
import fr.vsct.tock.bot.engine.action.Action
import fr.vsct.tock.shared.error
import mu.KotlinLogging

/**
 * A Story is a small unit of conversation about a specific topic.
 * It is linked to at least one intent - the [starterIntent].
 */
data class Story(
    val definition: StoryDefinition,
    val starterIntent: Intent,
    var currentStep: String? = null,
    val actions: MutableList<Action> = mutableListOf()
) {

    companion object {
        private val logger = KotlinLogging.logger {}
    }

    val lastAction: Action? get() = actions.lastOrNull()

    fun findCurrentStep() = currentStep?.let { s -> definition.steps.firstOrNull { it.name == s } }

    private fun StoryHandler.sendStartEvent(bus: BotBus): Boolean {
        //stops immediately if any startAction returns false
        return BotRepository.storyHandlerListeners.all {
            try {
                it.startAction(bus, this)
            } catch (throwable: Throwable) {
                logger.error(throwable)
                true
            }
        }
    }

    private fun StoryHandler.sendEndEvent(bus: BotBus) {
        BotRepository.storyHandlerListeners.forEach {
            try {
                it.endAction(bus, this)
            } catch (throwable: Throwable) {
                logger.error(throwable)
            }
        }
    }

    /**
     * Handles a request.
     */
    fun handle(bus: BotBus) {
        definition.storyHandler.apply {
            try {
                if (sendStartEvent(bus)) {
                    handle(bus)
                }
            } finally {
                sendEndEvent(bus)
            }
        }
    }

    /**
     * What is the probability of this request support by the current bot?
     */
    fun support(bus: BotBus): Double = definition.storyHandler.support(bus)

}

File: bot/admin/web/src/app/bot/story-builder/model/Story.ts
/*
 * Copyright (C) 2017/2020 e-voyageurs technologies
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


import {UserInteraction} from './UserInteraction';
import {InteractionDialog} from './InteractionDialog';

export class Story {
  name: string;
  rootUserInteraction: UserInteraction;

  constructor(name: string) {
    this.name = name;
  }

  setRootUserInteraction: (rootUserInteraction: UserInteraction) => void = (rootUserInteraction: UserInteraction) => {
    this.rootUserInteraction = rootUserInteraction;
  }

  getIntent: () => string = () => {
    if (!this.rootUserInteraction) {
      return 'unknown';
    }
    return this.rootUserInteraction.intent;
  }

  getSelectedInteractionDialogs: () => InteractionDialog[] = () => {
    const dialogs: InteractionDialog[] = [];
    if (this.rootUserInteraction) {
      let userInteraction = this.rootUserInteraction;
      let botInteraction = userInteraction.botInteraction;
      dialogs.push(new InteractionDialog(userInteraction, botInteraction));
      while (botInteraction.hasSelectedUserInteraction()) {
        userInteraction = botInteraction.getSelectedUserInteraction();
        botInteraction = userInteraction.botInteraction;
        dialogs.push(new InteractionDialog(userInteraction, botInteraction));
      }
    }
    return dialogs;
  }
}
